### ISSUE TITLE
Sanic 19.12.2 breaks sanic-cors 0.10.0.b1

### ISSUE THREAD

**Issue by nlnjnj** [ashleysommer/sanic-cors#41]
![071ECEAA-4DF8-4049-911F-E2F8534FF2D6](https://user-images.githubusercontent.com/5831940/74742207-25732b00-5299-11ea-9002-c7cff7fa75bf.png)
![552E7BD4-8954-446e-93C6-7F286EBD7C40](https://user-images.githubusercontent.com/5831940/74742209-26a45800-5299-11ea-9d73-131d30a13b3b.png)

    try:
        request_context = context.request[id(req)]
    except AttributeError:
        debug("Cannot find the request context. Is request already finished?")
        return False

in Sanic 19.12.2 this code with get an AttributeError and print debug message
which breaks the cors plugin

**Comment by ashleysommer** [comment]
Hi @nlnjnj 
Thanks for this report.

Do you know exactly what changed in Sanic 19.12.2 that causes this error?

**Comment by ashleysommer** [comment]
Hi again @nlnjnj 

I'm having some trouble understanding what your issue is here.

The case that AttributeError is thrown is already handled in the code.

Your question did prompt me to look thoroughly through the code and I found a different error, that part of the code should also check for LookupError in addition to AttributeErrror. I've pushed that fix into v0.10.0 release. It might fix the problem you are seeing.

**Comment by nlnjnj** [comment]
> Hi again @nlnjnj
> 
> I'm having some trouble understanding what your issue is here.
> 
> The case that AttributeError is thrown is already handled in the code.
> 
> Your question did prompt me to look thoroughly through the code and I found a different error, that part of the code should also check for LookupError in addition to AttributeErrror. I've pushed that fix into v0.10.0 release. It might fix the problem you are seeing.

The case that AttributeError is thrown is already handled in the code. but not handled correctly.
the request context cannot be found, but the request not finished.

**Comment by ashleysommer** [comment]
Hi @nlnjnj 
The only change between Sanic 19.12.0 and 19.12.2 is a missing __slots__ entry, and some documenation.
See here:
https://github.com/huge-success/sanic/compare/v19.12.0...v19.12.2
There's nothing different that would be causing the issue you are seeing.

I've run the full sanic-cors test suite against Sanic 19.12.2 and am not seeing any problems.

Are you able to please provide a reproducible example where the request context is not found, but the request is not finished? The Sanic-CORS logic (logic inherited from Sanic-Plugins-Framework) is that if a request context does not exist, then the request _must be_ finished.

**Comment by ashleysommer** [comment]
@nlnjnj 
Are you using sanic-cors on a Sanic Blueprint?
If so, I think I've tracked down a problem with the changes to blueprint handling in Sanic 19.12.0, its not compatible with the request context in Sanic-Plugins-Framework. I'm seeing it now in one of my own projects.

**Comment by ashleysommer** [comment]
@nlnjnj I've determined the cause of the problem. It is not related to Blueprints.

The is error is occuring when the Sanic router cannot find a valid handler for the given route, so throws a NotFound error. The Sanic-CORS exception handler can normally catch that error and apply the correct CORS headers to the error response.

But in Sanic 19.12.0 there was a change to the order of operations in the handler. The router now runs _before_ the request-middleware. So the sanic-cors request context has not yet set up when a router NotFound error is thrown.

I'm working on a workaround for this now.

**Comment by nlnjnj** [comment]
the issue may be related to Sanic-Plugins-Framework, with the following configuration:
      - sanic>=19.12.2
      - sanic-cors>=0.10.0.post2
      - sanic-plugins-framework==0.9.1

I still can not get cors work, and keeps seeing this debug message

> [DEBUG] CORS: CORS was handled in the exception handler, skipping

**Comment by FFX01** [comment]
I'm having the same issue as @nlnjnj . I'm using `0.10.0.post2`

**Comment by ashleysommer** [comment]
Hi @nlnjnj and @FFX01 
I've been trying to reproduce this problem without success.

Can you please run this simple snippet to verify that Sanic-CORS is working as expected in your environment?

```python3
from sanic import Sanic, __version__ as sanic_version
from sanic.exceptions import Unauthorized
from sanic.response import text
from sanic_cors import CORS, __version__ as cors_version
from spf import SanicPluginsFramework, __version__ as spf_version

assert (sanic_version == "19.12.2"), "Checking Sanic Version"
assert (cors_version == "0.10.0.post2"), "Checking Sanic-CORS version"
assert (spf_version == "0.9.1"), "Checking SPF version"

app = Sanic(__name__)
app.debug = True

spf = SanicPluginsFramework(app)

spf.register_plugin(CORS, automatic_options=True)

@app.route("/testme401")
def testme401(request):
    raise Unauthorized("testme401")

@app.route("/testme200")
def testme200(request):
    return text("testme200")

@app.route("/testme500")
def testme500(request):
    raise RuntimeError("testme500")

if __name__ == "__main__":
    client = app.test_client
    req, resp1 = client.get("/testme401")
    assert (resp1.status == 401), "Testing response code 401"
    assert (resp1.headers.get('access-control-allow-origin') == "*")
    req, resp2 = client.get("/testme404")
    assert (resp2.status == 404), "Testing response code 404"
    assert (resp2.headers.get('access-control-allow-origin') == "*")
    req, resp3 = client.get("/testme200")
    assert (resp3.status == 200), "Testing response code 200"
    assert (resp3.headers.get('access-control-allow-origin') == "*")
    req, resp4 = client.get("/testme500")
    # CORS on 500 error because intercept_exceptions is on by default
    assert (resp4.status == 500), "Testing response code 500"
    assert (resp4.headers.get('access-control-allow-origin') == "*")

```

> keeps seeing this debug message
> >  [DEBUG] CORS: CORS was handled in the exception handler, skipping

This debug message is normal, it means the Sanic Exception handler is working correctly, it is a signal to the CORS response-middleware that CORS is already applied by the CORS exception handler, so does not need to be applied again.

**Comment by nlnjnj** [comment]
> Hi @nlnjnj and @FFX01
> I've been trying to reproduce this problem without success.
> 
> Can you please run this simple snippet to verify that Sanic-CORS is working as expected in your environment?
> 
> ```python
> from sanic import Sanic, __version__ as sanic_version
> from sanic.exceptions import Unauthorized
> from sanic.response import text
> from sanic_cors import CORS, __version__ as cors_version
> from spf import SanicPluginsFramework, __version__ as spf_version
> 
> assert (sanic_version == "19.12.2"), "Checking Sanic Version"
> assert (cors_version == "0.10.0.post2"), "Checking Sanic-CORS version"
> assert (spf_version == "0.9.1"), "Checking SPF version"
> 
> app = Sanic(__name__)
> app.debug = True
> 
> spf = SanicPluginsFramework(app)
> 
> spf.register_plugin(CORS, automatic_options=True)
> 
> @app.route("/testme401")
> def testme401(request):
>     raise Unauthorized("testme401")
> 
> @app.route("/testme200")
> def testme200(request):
>     return text("testme200")
> 
> @app.route("/testme500")
> def testme500(request):
>     raise RuntimeError("testme500")
> 
> if __name__ == "__main__":
>     client = app.test_client
>     req, resp1 = client.get("/testme401")
>     assert (resp1.status == 401), "Testing response code 401"
>     assert (resp1.headers.get('access-control-allow-origin') == "*")
>     req, resp2 = client.get("/testme404")
>     assert (resp2.status == 404), "Testing response code 404"
>     assert (resp2.headers.get('access-control-allow-origin') == "*")
>     req, resp3 = client.get("/testme200")
>     assert (resp3.status == 200), "Testing response code 200"
>     assert (resp3.headers.get('access-control-allow-origin') == "*")
>     req, resp4 = client.get("/testme500")
>     # CORS on 500 error because intercept_exceptions is on by default
>     assert (resp4.status == 500), "Testing response code 500"
>     assert (resp4.headers.get('access-control-allow-origin') == "*")
> ```
> 
> > keeps seeing this debug message
> > > [DEBUG] CORS: CORS was handled in the exception handler, skipping
> 
> This debug message is normal, it means the Sanic Exception handler is working correctly, it is a signal to the CORS response-middleware that CORS is already applied by the CORS exception handler, so does not need to be applied again.

from sanic import Blueprint
from sanic import Sanic, __version__ as sanic_version
from sanic.response import text
from sanic_cors import CORS, __version__ as cors_version
from spf import __version__ as spf_version

assert (sanic_version == '19.12.2'), 'Checking Sanic Version'
assert (cors_version == '0.10.0.post2'), 'Checking Sanic-CORS version'
assert (spf_version == '0.9.1'), 'Checking SPF version'

app = Sanic(__name__)
app.debug = True

CORS(app, automatic_options=True)

blueprint = Blueprint('test')


@blueprint.get('/testme200')
def testme200(request):
    return text('/testme200')


blueprints = Blueprint.group(blueprint)

app.blueprint(blueprints)

if __name__ == '__main__':
    client = app.test_client
    req, resp1 = client.options('/testme200')

    req, resp3 = client.get('/testme200')
    assert (resp3.status == 200), 'Testing response code 200'
    assert (resp3.headers.get('access-control-allow-origin') == '*')

**Comment by ashleysommer** [comment]
Hi @nlnjnj 
> sanic.exceptions.MethodNotSupported: Method OPTIONS not allowed for URL /testme200

This is a different error than you described in previous comments. But we can discuss it here in this issue thread.

I see now this is another CORS incompatibility introduced with Sanic 19.12, that is  `automatic_options` doesn't work quite the same anymore, because the router is run before the Sanic-CORS request middleware.

You can workaround this for now by changing `@blueprint.get("/testme")` to `@blueprint.route("/testme", methods=("GET", "OPTIONS"))`

I will have a look at how hard it will be to patch this feature back working.

**Comment by ashleysommer** [comment]
@nlnjnj 
New patch release version 0.10.0.post3 pushed to address the "Method OPTIONS not allowed" error when using `automatic_options=True`. Please upgrade to see if it has fixed that issue for you.

**Comment by nlnjnj** [comment]
it did fix that issue,

**Comment by blutooth** [comment]
@ashleysommer Had the same issue, gonna test the patched version or downgrade sanic. Thanks for fixing things like this.

**Comment by ashleysommer** [comment]
Hi @blutooth 
When you say you had the same problem, do you mean the original problem as reported at the top of this thread where CORS is not applied in the exception handler (fixed in 0.10.0.post2) or the second issue where method options is not allowed when automatic_options is used (fixed in 0.10.0.post3)?

**Comment by ashleysommer** [comment]
Hi @nlnjnj @blutooth 
Looks like version 0.10.0.post3 didn't get published properly through to PyPI, this was because I forgot to run some tests before pushing it out.
Fixed the release and republished it, so v0.10.0.post3 with the latest bugfix should appear in pypi soon.
